### PR TITLE
Service now plugin

### PR DESCRIPTION
**Pull request by shiva-cirus**
Initial Release for Engineering Team Review

### CODE REVIEW

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    }
+  }
+
+  private void fetchData() {
+    tableName = split.getTableName();
+    tableNameField = pluginConf.getTableNameField();
+
+    ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(pluginConf);
+
+    //Get the table data
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java:137]: Nit: space after //. Similar comment for other comments in this file.

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    </pluginRepository>
+  </pluginRepositories>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api</artifactId>
+      <version>${cdap.version}</version>
+    </dependency>
+    <!--
```
Comment (rmstar) [inline pom.xml:71]: please remove commented out dependency if it's not required.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+  public String getPassword() {
+    return password;
+  }
+
+  public SourceValueType getValueType(FailureCollector collector) {
+    SourceValueType type = getValueType();
+    if (type != null) {
+      return type;
+    }
+
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java:232]: We need to handle the case where value type is a macro.

File: src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  private Schema createSchema(ServiceNowColumn column) {
+    switch (column.getTypeName().toLowerCase()) {
+      case "decimal":
+        return Schema.of(Schema.Type.DOUBLE);
+      case "integer":
+        return Schema.of(Schema.Type.INT);
+      case "boolean":
+        return Schema.of(Schema.Type.BOOLEAN);
+      case "reference":
+        /*
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java]: Not required

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    for (ServiceNowTableInfo tableInfo : tableInfos) {
+      String tableName = tableInfo.getTableName();
+      int totalRecords = tableInfo.getRecordCount();
+      if (totalRecords < PAGE_SIZE) {
+        //add single split for table and continue
+        resultSplits.add(new ServiceNowInputSplit(tableName, 0, totalRecords));
+        continue;
+      }
+
+      int pages = (tableInfo.getRecordCount() / PAGE_SIZE) + 1;
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    switch (fieldType) {
+      case STRING:
+        return convertToStringValue(fieldValue);
+      case DOUBLE:
+        return convertToDoubleValue(fieldValue);
+      case INT:
+        return convertToIntegerValue(fieldValue);
+      case BOOLEAN:
+        return convertToBooleanValue(fieldValue);
+      //case RECORD:
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Removed commented code as it is not required

File: src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+public class ServiceNowTableAPIClientImpl extends RestAPIClient {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowTableAPIClientImpl.class);
+  private static final String DATE_RANGE_TEMPLATE = "%sBETWEENjavascript:gs.dateGenerate('%s','start')" +
+    "@javascript:gs.dateGenerate('%s','end')";
+  private static final String TABLE_SYS_DICTIONARY = "sys_dictionary";
+  private static final String FIELD_CREATED_ON = "sys_created_on";
+  private static final String FIELD_UPDATED_ON = "sys_updated_on";
+  private static final String HEADER_NAME_TOTAL_COUNT = "X-Total-Count";
+  private static final String OAUTH_URL_TEMPLATE = "%s/oauth_token.do";
+  private static final boolean FETCH_SCHEMA_FROM_SYS_TABLE = false;
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/apiclient/ServiceNowTableAPIClientImpl.java]: No. Removed this code as it is not required.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return pos / (float) split.getLength();
+  }
+
+  @Override
+  public void close() throws IOException {
+    SQLException exception = null;
+
+    if (exception != null) {
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Removed this code as it is not required.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+
+  private Boolean convertToBooleanValue(Object fieldValue) {
+    if (fieldValue instanceof String && Strings.isNullOrEmpty(String.valueOf(fieldValue))) {
+      return null;
+    }
+
+    return Boolean.parseBoolean(String.valueOf(fieldValue));
+  }
+
+  private StructuredRecord convertToRecordValue(String fieldName, Schema fieldSchema, Object fieldValue) {
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Removed unused code as it is not required.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+   * Configure the input format to read tables from ServiceNow. Should be called from the mapreduce client.
+   *
+   * @param jobConfig the job configuration
+   * @param mode
+   * @param conf      the database conf
+   * @return Collection of ServiceNowTableInfo containing table and schema.
+   */
+  public static List<ServiceNowTableInfo> setInput(Configuration jobConfig, SourceQueryMode mode,
+                                                   ServiceNowSourceConfig conf) {
+    ServiceNowJobConfiguration jobConf = new ServiceNowJobConfiguration(jobConfig);
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java:59]: jobConf is used in getSplits() method.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+
+      int pages = (tableInfo.getRecordCount() / PAGE_SIZE) + 1;
+      int offset = 0;
+      int recordsOnPage = PAGE_SIZE;
+
+      for (int page = 1; page <= pages; page++) {
+        if (page == pages) {
+          recordsOnPage = totalRecords - offset;
+        }
+        resultSplits.add(new ServiceNowInputSplit(tableName, offset, recordsOnPage));
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java]: Done.

File: src/main/java/io/cdap/plugin/servicenow/source/util/SourceApplication.java [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+/**
+ * Indicates application for which data to be fetched from multiple tables in ServiceNow.
+ */
+public enum SourceApplication {
+
+  /**
+   * Data from tables related to Contract Management will be fetched
+   */
+  CONTRACT_MANAGEMENT("Contract Management", Arrays.asList("clm_m2m_contract_asset", "clm_m2m_rate_card_asset",
```
[the remark this answers is not in the document]
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/util/SourceApplication.java:33]: Added links to Readme file.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java (hunk 1 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java]: Done

File: pom.xml (hunk 2 above)
Comment (amolkokane76) [reply pom.xml]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java (hunk 3 above)
Comment (amolkokane76) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java:266]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+  private Schema createSchema(ServiceNowColumn column) {
+    switch (column.getTypeName().toLowerCase()) {
+      case "decimal":
+        return Schema.of(Schema.Type.DOUBLE);
+      case "integer":
+        return Schema.of(Schema.Type.INT);
+      case "boolean":
+        return Schema.of(Schema.Type.BOOLEAN);
+      case "reference":
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java:75]: Can we add a Data Types Mapping section in ServiceNow-batchsource.md similar to https://github.com/data-integrations/database-plugins/blob/develop/mysql-plugin/docs/Mysql-batchsource.md#data-types-mapping?

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+  public void testQueryModeInvalid() {
+    ServiceNowSourceConfig config = ServiceNowSourceConfigHelper.newConfigBuilder()
+      .setQueryMode(null)
+      .build();
+
+    try {
+      MockFailureCollector collector = new MockFailureCollector();
+      config.getQueryMode(collector);
+    } catch (ValidationException e) {
+      Assert.assertEquals(1, e.getFailures().size());
```
Comment (rmstar) [inline src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java:92]: This test case will pass (incorrectly) if config.getQueryMode() does not throw ValidationException when query mode is invalid. Similar comment for other test cases that expect ValidationException to be thrown.

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (rmstar) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java:59]: getSplits() is using a different ServiceNowJobConfiguration  instance. In setInput, we're creating a ServiceNowJobConfiguration instance, and don't seem to be doing anything with it.

File: src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java (hunk 13 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/servicenow/source/util/SchemaBuilder.java:75]: Done

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java (hunk 14 above)
Comment (amolkokane-cirus) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java:92]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java:59]: Even though getSplits is using different instance of ServiceNowJobConfiguration, in setInput when we are calling set methods of ServiceNowJobConfiguration instance, it is internally updating the jobConfig by reference. The value which is changed by reference is then read in getSplits.

File: src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java (hunk 14 above)
Comment (rmstar) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java:92]: How was this addressed?

Comment (amolkokane-cirus) [reply src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java]: We are asserting size of validation failures in collector outside the catch block

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+  private String user;
+
+  @Name(ServiceNowConstants.PROPERTY_PASSWORD)
+  @Macro
+  @Description("The password for ServiceNow Instance.")
+  private String password;
+
+  @Name(ServiceNowConstants.PROPERTY_VALUE_TYPE)
+  @Macro
+  @Nullable
```
[the remark this answers is not in the document]
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java]: Done

File: src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/servicenow/source/ServiceNowInputFormat.java:59]: Done

File: pom.xml [hunk 16]
```diff
@@ -0,0 +1,330 @@
...
+      <id>sonatype</id>
+      <url>https://oss.sonatype.org/content/groups/public</url>
+    </pluginRepository>
+  </pluginRepositories>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-etl-api</artifactId>
+      <version>${cdap.version}</version>
```
Comment (rmstar) [inline pom.xml:69]: scope should be provided
https://github.com/data-integrations/google-cloud/blob/develop/pom.xml#L148

File: pom.xml [hunk 17]
```diff
@@ -0,0 +1,330 @@
...
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>cdap-data-pipeline</artifactId>
+      <version>${cdap.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.cdap.cdap</groupId>
+      <artifactId>hydrator-test</artifactId>
+      <version>${cdap.version}</version>
+      <scope>provided</scope>
```
Comment (rmstar) [inline pom.xml:86]: should be test scope. https://github.com/data-integrations/google-cloud/blob/develop/pom.xml#L416

File: pom.xml (hunk 16 above)
Comment (amolkokane-cirus) [reply pom.xml:69]: Done

File: pom.xml (hunk 17 above)
Comment (amolkokane-cirus) [reply pom.xml]: Done

**Review by rmstar — APPROVED** [review]
Please squash commits.

**Comment by rmstar** [comment]
Squashed and merged in #2. Closing this PR.

**Pull request closed without merging**